app/api-client.js

```diff
@@ -1,170 +1,291 @@
+    if (!['readonly', 'readwrite'].includes(access_level)) {
+      throw new Error(`Invalid access_level "${access_level}"`);
+    }
+
+    let bucket_id = bucket;
 
-  api.list_buckets = function () {
-    return api_request({ endpoint: 's3buckets' });
-  };
+    if (typeof bucket === 'function' && bucket.prototype.constructor === Bucket) {
+      bucket_id = bucket.id;
+    }
+
+    return new AppS3Bucket({
+      app: this.id,
+      s3bucket: bucket_id,
+      access_level,
+    }).create();
+  }
+}
 
-  api.list_app_buckets = function (app_id) {
-    return api_request({ endpoint: `apps/${app_id}/s3buckets` });
-  };
+exports.App = App;
 
-  api.list_user_buckets = function (user_id) {
-    return api_request({ endpoint: `users/${user_id}/s3buckets` });
-  };
+class Bucket extends Model {
+  static get endpoint() {
+    return 's3buckets';
+  }
 
-  api.get_bucket = function (id) {
-    return api_request({ endpoint: `s3buckets/${id}` });
-  };
+  get apps() {
+    return new ModelSet(App, this.data.apps3buckets.map(as => as.app));
+  }
 
-  api.add_bucket = function (bucket) {
-    return api_request({ method: 'POST', endpoint: 's3buckets', resource: bucket });
-  };
+  get apps3buckets() {
+    return new ModelSet(AppS3Bucket, this.data.apps3buckets);
+  }
 
-  api.buckets = {
-    list: api.list_buckets,
-    get: api.get_bucket,
-    add: api.add_bucket,
-  };
+  get users() {
+    return new ModelSet(User, this.data.users3buckets.map(us => us.user));
+  }
 
-  api.add_apps3bucket = (apps3bucket) => {
-    return api_request({ method: 'POST', endpoint: 'apps3buckets', resource: apps3bucket });
-  };
+  get users3buckets() {
+    return new ModelSet(UserS3Bucket, this.data.users3buckets);
+  }
+}
 
-  api.update_apps3bucket = (apps3bucket) => {
-    return api_request({ method: 'PATCH', endpoint: `apps3buckets/${apps3bucket.id}`, resource: apps3bucket });
-  };
+exports.Bucket = Bucket;
 
-  api.delete_apps3bucket = (apps3bucket_id) => {
-    return api_request({ method: 'DELETE', endpoint: `apps3buckets/${apps3bucket_id}` });
-  };
 
-  api.apps3buckets = {
-    add: api.add_apps3bucket,
-    update: api.update_apps3bucket,
-    delete: api.delete_apps3bucket,
-  };
-
-  api.add_users3bucket = (users3bucket) => {
-    return api_request({ method: 'POST', endpoint: 'users3buckets', resource: users3bucket });
-  };
-
-  api.update_users3bucket = (users3bucket) => {
-    return api_request({ method: 'PATCH', endpoint: `users3buckets/${users3bucket.id}`, resource: users3bucket });
-  };
-
-  api.delete_users3bucket = (users3bucket_id) => {
-    return api_request({ method: 'DELETE', endpoint: `users3buckets/${users3bucket_id}` });
-  };
-
-  api.users3buckets = {
-    add: api.add_users3bucket,
-    update: api.update_users3bucket,
-    delete: api.delete_users3bucket,
-  };
-
-  return api;
-})();
+api.connect_bucket_to_app = apps3bucket => api.post('apps3buckets', body = apps3bucket);
```

Comment:
There is something unsettling about this line of code (it may be just me being tired!).

I think as the `api.post()` has side effect and it's not just mapping parenthesis and curly braces may help (I think Airbnb coding standard says something about it).

Also, I was about to say something about this function name but it seems like it's not used anywhere so we may just delete this line.

Reply:
Yep, that's right - it should have curly braces - and it's not used anywhere. I will delete it.